### PR TITLE
Generate CSR using wasm module

### PR DESCRIPTION
**Pull request by baluchicken**
## Description

This PR generates a rsa keypair inside a kernel using bearSSL. The stored key will then be used to generate a certificate signing request using a wasm module. (bearSSL at the moment does not support this). This CSR will then be used in an agent code(e.g.: istio controlplane) to generate a proper certificate.

The used module can be found: https://github.com/cisco-open/wasm-kernel-module-cli/tree/main/samples/csr-rust

## Type of Change

- [ ] Bug Fix
- [X] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [ ] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [ ] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [ ] All code style checks pass
- [ ] New code contribution is covered by automated tests
- [ ] All new and existing tests pass

### CODE REVIEW

File: rsa_tools.c [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+// Generates a 2048 bit long rsa key pair
+uint32_t generate_rsa_keys(br_rsa_private_key *rsa_priv, br_rsa_public_key *rsa_pub)
+{
+    br_rsa_keygen rsa_keygen = br_rsa_keygen_get_default();
+
+    unsigned char raw_priv_key[BR_RSA_KBUF_PRIV_SIZE(2048)];
+    unsigned char raw_pub_key[BR_RSA_KBUF_PUB_SIZE(2048)];
+
+
+    uint32_t result = rsa_keygen(&hmac_drbg_ctx.vtable, rsa_priv, raw_priv_key, rsa_pub, raw_pub_key, 2048, 3);
```
Comment (bonifaido) [inline rsa_tools.c:32]: can be an inline return

File: rsa_tools.c [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// If the der parameter is set to NULL then it computes only the length
+size_t encode_rsa_priv_key_to_der(unsigned char *der, br_rsa_private_key *rsa_priv, br_rsa_public_key *rsa_pub)
+{
+    br_rsa_compute_privexp rsa_priv_exp_comp = br_rsa_compute_privexp_get_default();
+    unsigned char priv_exponent[256];
+    size_t priv_exponent_size = rsa_priv_exp_comp(priv_exponent, rsa_priv, 3);
+    if (rsa_pub->nlen != priv_exponent_size) {
+        printk("Error happened during priv_exponent generation");
+    }
+    size_t len = br_encode_rsa_pkcs8_der(der, rsa_priv, rsa_pub, priv_exponent, priv_exponent_size);
```
Comment (bonifaido) [inline rsa_tools.c:47]: inline return

File: socket.c [hunk 3]
```diff
@@ -194,6 +201,10 @@ static wasm_socket_context *new_server_wasm_socket_context(proxywasm *p)
 {
 	wasm_socket_context *c = kzalloc(sizeof(wasm_socket_context), GFP_KERNEL);
 	c->sc = kmalloc(sizeof(br_ssl_server_context), GFP_KERNEL);
+	c->rsa_priv = kzalloc(sizeof(br_rsa_private_key), GFP_KERNEL);
```
Comment (bonifaido) [inline socket.c:204]: Are these new allocations freed anywhere?

**Review by bonifaido — CHANGES_REQUESTED** [review]

File: socket.c (hunk 3 above)
Comment (baluchicken) [reply socket.c:204]: Yes:) now they are freed `free_wasm_socket_context` function.

File: rsa_tools.c [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+{
+
+    br_prng_seeder seeder;
+
+    seeder = br_prng_seeder_system(NULL);
+
+    br_hmac_drbg_init(&hmac_drbg_ctx, &br_sha256_vtable, NULL, 0);
+    if (!seeder(&hmac_drbg_ctx.vtable))
+    {
+        printk(KERN_ERR "system source of randomness failed");
```
Comment (bonifaido) [inline rsa_tools.c:19]: This is a fatal error, so we should call `BUG()` here, see `assert.h`.

File: socket.c [hunk 5]
```diff
@@ -239,9 +250,19 @@ static void free_wasm_socket_context(wasm_socket_context *sc)
...
+		{
 			kfree(sc->sc);
+			kfree(sc->rsa_priv);
+			kfree(sc->rsa_pub);
+			kfree(sc->cert);
+		}
 		else
+		{
 			kfree(sc->cc);
+			kfree(sc->rsa_priv);
```
Comment (bonifaido) [inline socket.c:262]: Common parts could be moved out from the boolean arms.

**Review by bonifaido — APPROVED** [review]

**Review by bonifaido — APPROVED** [review]

**Pull request merged**